Fixed validation check when clicking password show/hide button.

nakagawamai opened the pull request:
# [課題No26](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#26)
パスワードの表示・非表示ボタンクリック時に、バリデーションチェックが走っていたため、
走らないように修正しました。

<img width="500" alt="スクリーンショット 2023-03-10 8 55 02" src="https://user-images.githubusercontent.com/99260576/224350247-318e0077-5a7f-4bba-b539-096219f52289.png">

## [CodeSandBox](https://codesandbox.io/s/github/nakagawamai/js-lesson/tree/fix/lesson26/lesson26) 

## Reasons for correction
Users only want to show or hide their passwords, so if validation checks are run when they click, they will be stressed.

## [Corresponding comment](https://github.com/nakagawamai/js-lesson/pull/32#issuecomment-1439331242)
> [nits]
When the Show Password button is pressed, validation runs, but the user may just want to show or hide the password.
That's what I was taught in my last review.
https://github.com/haru-programming/JavaScript-lessons/pull/40#issuecomment-1413659541
> 
> Using relatedTarget, you can identify the blur destination of the input.
> [Reference](https://developer.mozilla.org/ja/docs/Web/API/MouseEvent/relatedTarget)
> 
> Here is my correction. -> https://github.com/haru-programming/JavaScript-lessons/pull/40#issuecomment-1425779703
> Since it's NITS, I'll leave it to you to fix!

## Target file
- login.js
- register.js

## Functions already implemented in the previous PR
```
ログイン画面(login.html)を作ります
ユーザー名とメールアドレスどちらも入れられるinputにしてください
パスワードの実装は25と同じ。
パスワード忘れてた方 forgotpassword.htmlを作って遷移できるようにしてください。ページ内の実装はしないで良いです。
会員登録はこちらから課題24,25で作ったregister.htmlへ遷移できる。

パスワードは仮に固定値としてN302aoe3とし、メールアドレス、ご自身で決めたものでいいです。
(パスワードは固定値N302aoe3)
ログインページに遷移したらトークンがlocalStorageにあるかどうか確認してください。
(トークンはログインでメールアドレス、パスワードがあっていれば発行されます(下記※1で検証した時です)。
通常はAPIで受け取りますが今回はダミーで固定値として作ります)
トークンがあれば、コンテンツ画面ページへ、なければログインページをそのまま出します。
ログインページ内は送信ボタン押下時に。

{
  email: "kejimorita@gmail.com", // <- なんでも良いです
  password: "N302aoe3" // 固定値
}
// ※これらはレビュー時にレビュワーに教えてあげるかPRに書いて示してください
というものがsubmitの引数として渡ってきたら、
Promise を使ってnameとpasswordをそれぞれの値のプリミティブ比較で合っているか検証してください(※1)。
(ここのPromise辺りの実装はなんでもいいです)
サーバー内で検証していることを想定しているので、awaitしている関数内で行ってください。
合っていたらresolve、違う場合 reject。

その返値にトークンを返して判定してください。
返値のトークンは[Chance](https://chancejs.com/mobile/apple_token.html)を使用し生成してください。

// 成功時
{ token: "fafae92rfjafa03", ok: true, code: 200 }

// 失敗時
{ ok: false, code: 401 }
それに応じて画面ページか失敗しましたページへ遷移
成功したらそのtokenを、ログイン画面内でlocalStorageに埋め込んでください。
```

nakagawamai committed:
fix: Fixed validation check behavior.

Validation was running when password show/hide was clicked, so I fixed it.

nakagawamai commented:
お手数ですが、以下の要件で動作確認お願いできればと思います。

##Target page
- ログイン画面　login.html
- 会員登録画面　register.html

## Check List
- パスワード入力欄にフォーカス後、パスワードの表示・非表示ボタンをクリックした際に、バリデーションチェックが走らないか
- パスワード入力欄にフォーカス後、パスワードの表示・非表示ボタン以外の場所をクリックした際に、バリデーションチェックが走るか

## How to Check
### パスワードの表示・非表示ボタンをクリックをクリックしたときの挙動の確認方法
1.パスワードの入力欄にフォーカスを当てる
2.パスワードの表示・非表示ボタンをクリックする
3.条件に一致していない入力内容でも、バリデーションが走らないか確認
4.別の場所をクリックするとバリデーションが走るか確認

### パスワードの表示・非表示ボタン以外をクリックしたときの挙動の確認方法
1.パスワードの入力欄にフォーカスを当てる
2.パスワードの表示・非表示ボタン以外の場所をクリックする
3.バリデーションが走るか確認

nakagawamai commented on lesson26/src/js/login.js:
```diff
@@ -87,3 +89,8 @@ for (const input of inputSelector){
 }
 
 togglePasswordButton.addEventListener('click', () => togglePassword(passwordInput,togglePasswordButton));
+
+const isRelatedTogglePassword = (event) => {
+    const related = event.relatedTarget ? event.relatedTarget.id : "unknown";
+    return related === "js-toggle-password";
+}
```
If you try to get the `event.relatedTarget.id` directly, it will not work, so this is how it is written.
 ref: https://developer.mozilla.org/ja/docs/Web/API/MouseEvent/relatedTarget#javascript

haru-programming reviewed (commented):
I have checked the operation first.🙇‍♀️

日本語でごめんなさい・・・

入力してくださいエラーが出ている時に、何かしら入力した状態で

① input外を押した時にバリデーションが通っていないようです
② 表示・非表示ボタンを押した時に「入力してください」の場合だけエラーが消えるとなお良いと思いました。が、こちらは完全にnits なので修正必要ないです🙇‍♀️

https://user-images.githubusercontent.com/67426438/224465405-a15f36ec-cc76-422c-9f7c-d5143eb2ba32.mov

nakagawamai commented:
@haru-programming 
> 入力してくださいエラーが出ている時に、何かしら入力した状態で
> 
> ① input外を押した時にバリデーションが通っていないようです
> ② 表示・非表示ボタンを押した時に「入力してください」の場合だけエラーが消えるとなお良いと思いました。が、こちらは> 完全にnits なので修正必要ないです🙇‍♀️

コメントありがとうございます！
inputでしかバリデーションチェックを走らせていないせいですね。。
表示・非表示ボタンをクリックした際にも、blurでバリデーションチェックが走るように、修正しようと思います！

nakagawamai committed:
refactor: Renamed function and added relatedTarget argument.

Added arguments for ease of reuse.

nakagawamai committed:
fix: Add blur event for toggle password button.

Fixed because validation was not running after clicking the display/hide button

nakagawamai committed:
refactor: Delete the error message after clicking the password display/hide button.

nakagawamai committed:
fix: Reflect validation check on register page.

nakagawamai commented on lesson26/src/js/login.js:
```diff
@@ -90,7 +90,7 @@ for (const input of inputSelector){
     const related = event.relatedTarget ? event.relatedTarget.id : "unknown";
-    return related === "js-toggle-password";
+    return related === target;
 }
```
Arguments have been added for ease of use.

nakagawamai commented on lesson26/src/js/login.js:
```diff
@@ -90,6 +90,13 @@ for (const input of inputSelector){
+    validation.checkPassword(passwordInput);
+    changeDisabledStatusSubmitButton();
+});
+
```
Added a blur event to the Show/Hide Password button so that validation checks can be run.

nakagawamai commented on lesson26/src/js/login.js:
```diff
@@ -88,7 +88,10 @@ for (const input of inputSelector){
+togglePasswordButton.addEventListener('click', () => {
+    validation.removeErrorMessage(passwordInput);
+    togglePassword(passwordInput,togglePasswordButton);
+});
```
Fixed to remove error messages in the password entry field when clicking the Show/Hide Password button.

nakagawamai commented:
@haru-programming 
お疲れ様です！下記、修正いたしました。https://github.com/nakagawamai/js-lesson/pull/34/commits/9f0670a3aa45c33f2d46154cc022aced60432298
> 入力してくださいエラーが出ている時に、何かしら入力した状態で
> 
> ① input外を押した時にバリデーションが通っていないようです


下記についても、よりユーザーの操作性がよくなると思ったので、修正しました！https://github.com/nakagawamai/js-lesson/pull/34/commits/4d14cca50eeca3990277fdbad378f5681f8cfaeb
> ② 表示・非表示ボタンを押した時に「入力してください」の場合だけエラーが消えるとなお良いと思いました。が、こちらは> 完全にnits なので修正必要ないです🙇‍♀️

丁寧にみてくださり、ありがとうございます。
お手隙の際に、ご確認お願いいたします。

wattttaw commented:
I have confirmed the above operation, so I approve.

sorry if there is any omission.
and i'm not a good english speaker.

haru-programming commented:
Q.
修正ありがとうございます！
> 入力してくださいエラーが出ている時に、何かしら入力した状態で
① input外を押した時にバリデーションが通っていないようです

OKだと思います！

> ② 表示・非表示ボタンを押した時に「入力してください」の場合だけエラーが消えるとなお良いと思いました。が、こちらは> 完全にnits なので修正必要ないです🙇‍♀️

こちらもOKだと思います！

しかし、バリデーションでエラーが出ている時に表示・非表示ボタンを押した時もエラーが消えてしまっているので、こちらは消えない方が良いと思いますがどうでしょうか？
→ 最初の問いである「パスワードの表示・非表示ボタンクリック時に、バリデーションチェックが走っている」に該当してしまうと思いました。

すみません、元々の仕様からかなり離れてしまっているため、後者の②の対応はお任せします🙇‍♀️
荒らしてしまってすみません🙇‍♀️

https://user-images.githubusercontent.com/67426438/224723294-05bacb60-0d62-4d37-a3c7-0cb783cc9866.mov

haru-programming reviewed (commented):
Q.
修正ありがとうございます！
> 入力してくださいエラーが出ている時に、何かしら入力した状態で
① input外を押した時にバリデーションが通っていないようです

OKだと思います！

> ② 表示・非表示ボタンを押した時に「入力してください」の場合だけエラーが消えるとなお良いと思いました。が、こちらは> 完全にnits なので修正必要ないです🙇‍♀️

こちらもOKだと思います！

しかし、バリデーションでエラーが出ている時に表示・非表示ボタンを押した時もエラーが消えてしまっているので、こちらは消えない方が良いと思いますがどうでしょうか？
→ 最初の問いである「パスワードの表示・非表示ボタンクリック時に、バリデーションチェックが走っている」に該当してしまうと思いました。

すみません、元々の仕様からかなり離れてしまっているため、後者の②の対応はお任せします🙇‍♀️
荒らしてしまってすみません🙇‍♀️

https://user-images.githubusercontent.com/67426438/224723294-05bacb60-0d62-4d37-a3c7-0cb783cc9866.mov

nakagawamai committed:
refactor: Fixed to leave an error message.

Deleted the process to erase the error message when the display/hide button is clicked.

nakagawamai commented:
@haru-programming 
> しかし、バリデーションでエラーが出ている時に表示・非表示ボタンを押した時もエラーが消えてしまっているので、こちら> > は消えない方が良いと思いますがどうでしょうか？
> → 最初の問いである「パスワードの表示・非表示ボタンクリック時に、バリデーションチェックが走っている」に該当してし> まうと思いました。

確かにその通りです！

空欄チェックのエラーメッセージ（入力してください）のみを消すことが、難しく、
エラーメッセージを一律で消すことしかできないため、いったん②の仕様は、実装からはずしました。

何度も確認していただきすみませんが、下記ご確認いただけると助かります！
https://github.com/nakagawamai/js-lesson/pull/34/commits/9ca970d94b91e3881bda218c62fc8a50ce4a335b

haru-programming commented on lesson26/src/js/login.js:
```diff
@@ -87,3 +89,15 @@ for (const input of inputSelector){
 
 togglePasswordButton.addEventListener('click', () => togglePassword(passwordInput,togglePasswordButton));
+
+togglePasswordButton.addEventListener('blur', (event) => {
+    if(isRelatedTarget(event,"current-password")) return;
+
+    validation.checkPassword(passwordInput);
+    changeDisabledStatusSubmitButton();
+});
```
```suggestion
togglePasswordButton.addEventListener('blur', (event) => {
    if(isRelatedTarget(event,"current-password")) return;

    validation.removeErrorMessage(passwordInput); //エラーをリセットしたい
    validation.checkPassword(passwordInput);
    changeDisabledStatusSubmitButton();
});
```
何度も申し訳ないです🙇‍♀️
①入力してくださいエラーが出ている状態で
②何かしら入力する
③表示・非表示ボタンを押す
④そのあとinput以外を押す
この時、入力してくださいエラーが出続けていて、バリデーションの内容がユーザーに伝わらないので、この点だけ直すと良いと思ったのですがどうでしょうか？

refactorは必要かもしれませんが、上記コードを書くなどしてエラーメッセージをリセットする必要があるかと思います。あるいは、リセットをcheckPassword(passwordInput)関数内で行うか（こちら未検証です🙇‍♀️）

何度も申し訳ありません。。。



https://user-images.githubusercontent.com/67426438/224870851-fc0a0203-5719-4e60-b5cc-9b4733a8014d.mov

nakagawamai replied:
具体的なコードをありがとうございます！
こちらこそ、何度も申し訳ありません。。
確かに古いエラーメッセージが残ってしまっているので、新しいエラー内容をユーザーに伝える必要がありました。

ご提案いただいたコードと同様のものに修正いたしましたので、
ご確認いただけると助かります。何度もすみません:pray:
https://github.com/nakagawamai/js-lesson/pull/34/commits/a12197e3c26a037be45cb0d36e7ab51144cafc2f

nakagawamai committed:
fix: Add delete error message.

Added removal of stale error messages to show/hide button blur events to better convey to the user what the current error is.

nakagawamai committed:
refactor: Change to one line.
